pennylane/compiler/python_compiler/transforms/quantum/tree_traversal.py

```diff
@@ -0,0 +1,2477 @@
+        """Unroll an scf.ForOp into separate branches for each operator."""
+
+        def find_constant_bound(bound: SSAValue) -> tuple[bool, Operation | None]:
+            """Find the constant value of a SSA in ForOp bounds"""
+
+            check_bound = bound
+            while True:
+                if isinstance(check_bound.owner, arith.ConstantOp):
+                    return True, check_bound.owner
+                if isinstance(check_bound.owner, stablehlo.ConstantOp):
+                    return True, check_bound.owner
+                if isinstance(check_bound, BlockArgument):
+                    return False, None
+                if len(check_bound.owner.operands) == 0:
+                    return False, None
+
+                check_bound = check_bound.owner.operands[0]
+
+        ub_found, ub_op = find_constant_bound(op.ub)
+        lb_found, lb_op = find_constant_bound(op.lb)
+        step_found, step_op = find_constant_bound(op.step)
+
+        assert (
+            lb_found and ub_found and step_found
+        ), "UnrollLoopPattern: Cannot unroll loop, bounds or step are not constant."
+
+        def check_extract_value(bound: Operation) -> int:
+
+            is_IntegerAttr: bool = isinstance(bound.value, builtin.IntegerAttr)
+            is_ElementsAttr: bool = isinstance(bound.value, builtin.DenseIntOrFPElementsAttr)
+
+            assert (
+                is_IntegerAttr or is_ElementsAttr
+            ), "UnrollLoopPattern: The ForOp bound should come from arith.ConstantOp or stablehlo.ConstantOp"
+
+            if is_IntegerAttr:
+                return bound.value.value.data
+
+            # is_ElementsAttr
+            return bound.value.data.data[0]
+
+        lb = check_extract_value(lb_op)
+        ub = check_extract_value(ub_op)
+        step = check_extract_value(step_op)
+
+        iter_args: tuple[SSAValue, ...] = op.iter_args
+
+        i_arg, *block_iter_args = op.body.block.args
+
+        inner_op_clone: Operation | None = None
+
+        for i in range(lb, ub, step):
+            i_op = rewriter.insert_op(
+                arith.ConstantOp(builtin.IntegerAttr(i, builtin.IndexType())),
+                InsertPoint.before(op),
+            )
+            i_op.result.name_hint = i_arg.name_hint
+
+            value_mapper: dict[SSAValue, SSAValue] = dict(
+                zip(block_iter_args, iter_args, strict=True)
+            )
+            value_mapper[i_arg] = i_op.result
+
+            for inner_op in op.body.block.ops:
+                if isinstance(inner_op, scf.YieldOp):
+                    iter_args = tuple(value_mapper.get(val, val) for val in inner_op.arguments)
+                else:
+                    inner_op_clone = inner_op.clone(value_mapper)
+                    rewriter.insert_op(inner_op_clone, InsertPoint.before(op))
+
+        if inner_op_clone is not None:
+            op.results[0].replace_by(inner_op_clone.results[0])
```

Comment:
This looks strange to me, a for loop could have multiple result values no? And why are they being replaced with the result of the last cloned op?